Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean

```diff
@@ -93,25 +93,133 @@ namespace CoordinateRing
+#align weierstrass_curve.coordinate_ring.smul_basis_eq_zero WeierstrassCurve.Affine.CoordinateRing.smul_basis_eq_zero
+
+lemma exists_smul_basis_eq (x : W.CoordinateRing) :
+    ∃ p q : R[X], p • (1 : W.CoordinateRing) + q • mk W Y = x := by
+  have h := (CoordinateRing.basis W).sum_equivFun x
+  erw [Fin.sum_univ_succ, Fin.sum_univ_one, basis_zero, basis_one] at h
+  exact ⟨_, _, h⟩
+#align weierstrass_curve.coordinate_ring.exists_smul_basis_eq WeierstrassCurve.Affine.CoordinateRing.exists_smul_basis_eq
+
+variable (W)
+
+lemma smul_basis_mul_C (y : R[X]) (p q : R[X]) :
+    (p • (1 : W.CoordinateRing) + q • mk W Y) * mk W (C y) =
+      (p * y) • (1 : W.CoordinateRing) + (q * y) • mk W Y := by
+  simp only [smul, _root_.map_mul]
+  ring1
+set_option linter.uppercaseLean3 false in
+#align weierstrass_curve.coordinate_ring.smul_basis_mul_C WeierstrassCurve.Affine.CoordinateRing.smul_basis_mul_C
+
+lemma smul_basis_mul_Y (p q : R[X]) :
+    (p • (1 : W.CoordinateRing) + q • mk W Y) * mk W Y =
+      (q * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)) • (1 : W.CoordinateRing) +
+        (p - q * (C W.a₁ * X + C W.a₃)) • mk W Y := by
+  have Y_sq : mk W Y ^ 2 =
+      mk W (C (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆) - C (C W.a₁ * X + C W.a₃) * Y) := by
+    exact AdjoinRoot.mk_eq_mk.mpr ⟨1, by rw [polynomial]; ring1⟩
+  simp_rw [smul, add_mul, mul_assoc, ← sq, Y_sq, C_sub, map_sub, C_mul, _root_.map_mul]
+  ring1
+set_option linter.uppercaseLean3 false in
+#align weierstrass_curve.coordinate_ring.smul_basis_mul_Y WeierstrassCurve.Affine.CoordinateRing.smul_basis_mul_Y
+
+section map
+
+variable {S} [CommRing S] (f : R →+* S)
+
+/-- Given a Weierstrass curve `W` over a ring `R`, a ring homomorphism `f : R →+* S` induces
+a homomorphism from the coordinate ring of `W` to the coordinate ring of the "base-changed"
+curve `W.map f`. -/
+noncomputable def map : W.CoordinateRing →+* CoordinateRing (W.map f) :=
+  AdjoinRoot.lift ((AdjoinRoot.of _).comp <| mapRingHom f) (AdjoinRoot.root _) <| by
+    simp [← eval₂_map, ← map_polynomial]
+
+variable {W f}
+
+lemma map_mk {x : R[X][X]} : map W f (mk W x) = mk (W.map f) (x.map <| mapRingHom f) := by
+  simp only [map, AdjoinRoot.lift_mk, ← eval₂_map]; apply AdjoinRoot.aeval_eq
+
+protected lemma map_smul {x : R[X]} {y : W.CoordinateRing} :
+    map W f (x • y) = x.map f • map W f y := by rw [smul, _root_.map_mul, map_mk, smul, map_C]; rfl
+
+open Function in
+lemma map_injective (hf : Injective f) : Injective (map W f) :=
+  (injective_iff_map_eq_zero _).mpr fun y hy ↦ by
+    obtain ⟨p, q, rfl⟩ := exists_smul_basis_eq y
+    simp_rw [map_add, CoordinateRing.map_smul, map_one, map_mk, map_X] at hy
+    obtain ⟨hp, hq⟩ := smul_basis_eq_zero hy
+    rw [Polynomial.map_eq_zero_iff hf] at hp hq
+    simp_rw [hp, hq, zero_smul, add_zero]
+
+instance instIsDomainCoordinateRing [IsDomain R] : IsDomain W.CoordinateRing :=
+  haveI : IsDomain (CoordinateRing <| W.map <| algebraMap R <| FractionRing R) :=
```

Comment:
I think that `have` in Lean 4 is `haveI` in Lean 3? Meanwhile `haveI` in Lean 4 inlines this, but that's not needed.

Reply:
Yeah, this is my bad habit, but probably harmless ...